### PR TITLE
Feature/create form

### PR DESCRIPTION
**Pull request by petia-latinova**

### CODE REVIEW

File: batch-nft-reveal/app/components/create/Create.tsx [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+  const [tokenNameEntered, setTokenNameEntered] = useState(false);
+  const [symbolEntered, setSymbolEntered] = useState(false);
+  const [maxSupplayEntered, setMaxSupplyEntered] = useState(false);
+  const [mintCostEntered, setMintCostEntered] = useState(false);
+  const [revealBatchSizeEntered, setRevealBatchSizeEntered] = useState(false);
+  const [revealIntervalEntered, setRevealIntervalEntered] = useState(false);
+  const [vrfCoordinatorV2Entered, setVrfCoordinatorV2Entered] = useState(false);
+  const [vrfSubscriptionIdEntered, setVrfSubscriptionIdEntered] = useState(false);
+  const [vrfGasLaneEntered, setVrfGasLaneEntered] = useState(false);
+  const [vrfCallbackGasLimitEntered, setVrfCallbackGasLimitEntered] = useState(false);
```
Comment (imollov) [inline batch-nft-reveal/app/components/create/Create.tsx:31]: All this state can be computed.

**Comment by imollov** [comment]
@petia-latinova the form must look like in the mockup. Currently, there are 1) missing fields, 2) unnecessary fields and 3) missing elements (divider, title).

File: batch-nft-reveal/app/components/create/Create.tsx [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+import { Error } from '../../components/Error'
+
+export function Create(): JSX.Element {
+  const { account } = useEthers()
+
+  const [hasError, setHasError] = useState(false);
+
+  const [tokenName, setTokenName] = useState("");
+  const [symbol, setSymbol] = useState("");
+  const [maxSupplay, setMaxSupply] = useState("");
```
Comment (imollov) [inline batch-nft-reveal/app/components/create/Create.tsx:13]:
```suggestion
  const [maxSupply, setMaxSupply] = useState("");
```

File: batch-nft-reveal/app/components/create/Create.tsx [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+            bgColor="white"
+            onChange={(event) => setTokenName(event.target.value)}
+            onKeyUp={() => setTokenNameEntered(true)}
+            onPaste={() => setTokenNameEntered(true)}
+          />
+        </Tooltip>
+        {isInvalidName && <FormErrorMessage>Name is not valid.</FormErrorMessage>}
+      </FormControl>
+
+      <FormControl isInvalid={symbolEntered && isInvalidSymbol}>
```
Comment (imollov) [inline batch-nft-reveal/app/components/create/Create.tsx:97]:
```suggestion
      <FormControl mt="4" isInvalid={symbolEntered && isInvalidSymbol}>
```

File: batch-nft-reveal/app/components/create/Create.tsx [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+  }
+
+  const isNotBytes32 = (value) => {
+    return !/^0x[a-fA-F0-9]{64}$/.test(value);
+  }
+
+  const isInvalidName = !tokenName;
+  const isInvalidSymbol = !symbol;
+  const isInvalidMaxSupplay = isNotNumber(maxSupplay) || Number(maxSupplay) <= 0;
+  const isInvalidMintCost = isNotNumber(mintCost); // TODO now is in satoshi... maybe make it fractional number
```
Comment (imollov) [inline batch-nft-reveal/app/components/create/Create.tsx:54]: There is no satoshi unit on Ethereum. The cost is in ETH, as you can see in the mockup.

**Comment by petia-latinova** [comment]
> @petia-latinova tests are failing, might be because `yarn.lock` is changed. Is there any reason for that?

The tests were failing and the versel deploy too and I deleted them and generated again to see if that will fix the problem. I removed them now. You are right, there is no reason to be changed.

**Comment by imollov** [comment]
@petia-latinova pls check this best practice and refactor: https://refactoring.com/catalog/removeDoubleNegative.html

**Comment by imollov** [comment]
@petia-latinova same as load page, submit button is not enabled when all fields are valid.

**Comment by imollov** [comment]
> @petia-latinova same as load page, submit button is not enabled when all fields are valid.

It's because there's no account connected. Maybe we should add a tooltip in that case?

File: batch-nft-reveal/app/components/create/Create.tsx [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    } else if (!value || !/^(0|([1-9][0-9]*))(\.[0-9]{0,18})?$/.test(value)) {
+      error = 'must be positive fractional number'
+    }
+    return error
+  }
+
+  return (
+    <form onSubmit={formik.handleSubmit}>
+      <Container maxW="mt">
+        <FormikProvider value={formik}>
```
Comment (imollov) [inline batch-nft-reveal/app/components/create/Create.tsx:98]: What's the purpose of this provider?

File: batch-nft-reveal/app/components/create/Create.tsx [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+      error = 'is required'
+    } else if (!value || !/^(0|([1-9][0-9]*))(\.[0-9]{0,18})?$/.test(value)) {
+      error = 'must be positive fractional number'
+    }
+    return error
+  }
+
+  return (
+    <form onSubmit={formik.handleSubmit}>
+      <Container maxW="mt">
```
Comment (imollov) [inline batch-nft-reveal/app/components/create/Create.tsx:97]: This should not be part of the form component.

File: batch-nft-reveal/app/components/create/Create.tsx [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+              bg="white"
+              name="mintCost"
+              validate={isNotFractualNumber}
+            />
+            {formik.errors.mintCost && formik.touched.mintCost && (
+              <FormErrorMessage>
+                Cost(ETH) {formik.errors.mintCost}
+              </FormErrorMessage>
+            )}
+            <span></span>
```
Comment (imollov) [inline batch-nft-reveal/app/components/create/Create.tsx:159]: ?

File: batch-nft-reveal/app/components/create/Create.tsx [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+              name="revealInterval"
+              validate={isNotNumber}
+            />
+            {formik.errors.revealInterval && formik.touched.revealInterval && (
+              <FormErrorMessage>
+                Interval(Seconds) {formik.errors.revealInterval}
+              </FormErrorMessage>
+            )}
+          </FormControl>
+          <Tooltip
```
Comment (imollov) [inline batch-nft-reveal/app/components/create/Create.tsx:246]: How it is:
Screenshot 2022-08-24 at 17 51 27

How it should be:
Screenshot 2022-08-24 at 17 53 30

File: batch-nft-reveal/app/components/create/Create.tsx (hunk 5 above)
Comment (petia-latinova) [reply batch-nft-reveal/app/components/create/Create.tsx]: Without `FormikProvider` the validations in formik will not be working.
`const formik = useFormik(...)` creates the state for formik with all it's functionality.
The provider connects the variable `formik` to the web code inside it.

There are other ways to use formik without `FormikProvider` but with the provider you can use other states and hooks which are not working with some of the other ways you can use formik.

**Pull request merged**